<h1>MNIST dataset trained using Keras</h1>
<p>Let us start with importing the libraries we will use. We import mnist, numpy, and we import models and layers for our neural network. We import to.categorical method to convert our labels into unit vectors for our training.</p>

In [1]:
from tensorflow.keras.datasets import mnist
import numpy as np
from keras import models 
from keras import layers
from keras.utils import to_categorical 
import matplotlib.pyplot as plt

Using TensorFlow backend.


<p>We load the data.</p>

In [2]:
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()
print(train_images.shape)

(60000, 28, 28)


<p>We create our model and add the layers. We will have 512 nodes in the hidden layer.</p>

In [3]:
model = models.Sequential() # Saying that the model is a sequential model
model.add(layers.Dense(512, activation='relu', input_shape=(28 * 28,))) # input shape is the image pixel number
model.add(layers.Dense(10, activation='softmax')) # first entry is the number of nodes, digits

<p>We specify the optimizer and the loss function and the metrics we will need.</p>

In [ ]:
model.compile(optimizer='rmsprop', 
    loss='categorical_crossentropy', 
    metrics=['accuracy'])

<p>Now we reshape our data so that each image becomes a column in the input matrix.</p>

In [5]:
train_images = train_images.reshape((60000, 28 * 28)) 
train_images = train_images.astype('float32') / 255

In [6]:
test_images = test_images.reshape((10000, 28 * 28)) 
test_images = test_images.astype('float32') / 255

In [7]:
train_labels = to_categorical(train_labels) 
test_labels = to_categorical(test_labels)

<p>Now we train the data for 25 epochs.</p>

In [8]:
train = model.fit(train_images, train_labels, epochs=25, batch_size=128)

Epoch 1/25
60000/60000 [==============================] - 2s 31us/step - loss: 0.2576 - accuracy: 0.9246
Epoch 2/25
60000/60000 [==============================] - 2s 29us/step - loss: 0.1051 - accuracy: 0.9682
Epoch 3/25
60000/60000 [==============================] - 2s 29us/step - loss: 0.0689 - accuracy: 0.9792
Epoch 4/25
60000/60000 [==============================] - 2s 31us/step - loss: 0.0502 - accuracy: 0.9850
Epoch 5/25
60000/60000 [==============================] - 2s 33us/step - loss: 0.0377 - accuracy: 0.9887
Epoch 6/25
60000/60000 [==============================] - 2s 33us/step - loss: 0.0289 - accuracy: 0.9912
Epoch 7/25
60000/60000 [==============================] - 2s 33us/step - loss: 0.0224 - accuracy: 0.9936
Epoch 8/25
60000/60000 [==============================] - 2s 33us/step - loss: 0.0170 - accuracy: 0.9951
Epoch 9/25
60000/60000 [==============================] - 2s 32us/step - loss: 0.0133 - accuracy: 0.9961
Epoch 10/25
60000/60000 [==============================

<p>We can check the training history, meaning that we check the loss and accuracy of each step. But frankly for our practical purposes, the training is done and now we can use the test data to test the neural network.</p>

In [13]:
train.history.keys()

dict_keys(['loss', 'accuracy'])

In [14]:
test_loss, test_acc = model.evaluate(test_images,test_labels)

10000/10000 [==============================] - 0s 31us/step


<p>Let us see how well our network did in the test data.</p>

In [18]:
print(test_acc)

0.9825999736785889


<p>We can see that we have simpyle 98% success with this network.</p>